### ISSUE TITLE
Paired / unpaired bracket handling is … eccentric

### ISSUE THREAD

**Issue by kopischke** [fish-shell/fish-shell#7970]
Running fish version 3.2.2 (installed via Homebrew) on macOS 11.3 in the stock Terminal.app. I tried fish without third-party customisations by executing `sh -c 'env HOME=$(mktemp -d) fish'` and checked whether it affected the behaviour I am reporting (it did not).

---------

One of the finer points of the fish syntax that gave me trouble while working on my syntax plugin for the [Nova editor](https://nova.app) was correctly handling unpaired pair characters. While quotes and parens are straightforward and braces are at least consistent (if less easily modelled in the syntax file), fish’s handling of paired / unpaired brackets I find frankly wonky.

Brackets are handled differently from other pair characters in that an opening bracket is valid as the first character of a shell word (presumably because forcing users to escape the `[` alias for `test` was considered a no-go?) but otherwise, nominally, brackets need to be paired to not trigger a syntax error. Actual parsing begs to differ:

1. Closing brackets are always valid, even when dangling (i.e. having no matching opening bracket); this is inconsistent with how parens and braces work.
2. A closing bracket in the same shell word will count as the closing pair match for any number of opening brackets: `echo foo[[[[[[bar]` does not trigger a syntax error. This is inconsistent with how all other paired characters work.
3. The cherry on the cake being that the closing bracket can be located _anywhere_ inside the shell word, even when its logical relationship to the opening one is tenuous, e.g. as part of index expansion: `echo foo[$var[i]` does not trigger a syntax error either.

**Comment by faho** [comment]
Duplicate of #2485

**Comment by kopischke** [comment]
I had seen this issue but missed the part where this specific issue is discussed. Apologies for the noise.

**Comment by kopischke** [comment]
@faho one follow-up question, if I may: is the behaviour I described considered a bug, to be changed in some future fish version? In find that a bit hard to gauge from [#2485](https://github.com/fish-shell/fish-shell/issues/2485#issuecomment-427373424), where it only crops up towards the end of the thread. I am trying to not implement fish quirks clearly considered bugs in my syntax extension, so I’d be grateful for a pointer.

**Comment by faho** [comment]
The exact behavior here would have to be nailed down. On the one hand it's nice if brackets are easy to use as literal characters - like `{}` braces, on the other its nice if the matching is intuitive.

Looking closer at this, I would expect these concrete examples to stay as they are.

The reason for that is that `[` only becomes special by being used with a variable expansion or command substitution, so the "grammar" is

```fish
$variable[index]
# or
(command)[index]
```

while just `word[index]` doesn't actually trigger any indexing, so there is no need for `[]` to be special.

In your examples:

- `foo[[[[[[bar]` has no `$`, so there's no indexing - this should just be literal
- `echo foo[$var[i]` only has indexing *on the second `[` and the closing `]`. So no indexing on the first `[`, so that one is literal.

**Comment by kopischke** [comment]
Thanks for your answer! I’m fine either way when it comes to my plugin (I had implemented the behaviour before starting to actually question it), and I wholeheartedly agree with the rationale. However, that sounds like opening brackets shouldn’t _ever_ be considered unpaired, unless they start indexing?

Amusingly, this is _exactly_ where I started conceptually, before noticing that this is not the current behaviour 😅.
